### PR TITLE
feat: order list filtering

### PR DESCRIPTION
**Pull request by dhhyi**
## PR Type

[x] Feature

## What Is the New Behavior?

Order List supports filtering by date range, product ID and order number.
If the number of orders exceeds the search result limit (=30) a 'load more' button is shown to fetch more orders from the server.

## Does this PR Introduce a Breaking Change?

[x] No

## Other Information

Requires ICM 11.7

[AB#92680](https://dev.azure.com/intershop-com/cefd1005-00a7-4c79-927f-a16947d1b2e6/_workitems/edit/92680)

### CODE REVIEW

**Review by SGrueber — COMMENTED** [review]
image

If there is no order at all the filter should be suppressed and it should look like before (see above)

If the search result is empty a different message should be shown, but the translation is missing for the key

image

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/assets/i18n/fr_FR.json [hunk 1]
```diff
@@ -210,11 +210,20 @@
...
   "account.order.most_recent.heading": "Les commandes les plus récentes",
   "account.order.questions.note": "Veuillez consulter la section <a href=\"{{0}}\">Aide</a> de notre site Web pour des renseignements détaillés sur votre commande et l’expédition ou <a href=\"{{1}}\">Contactez-nous</a> 24 heures sur 24.",
   "account.order.questions.title": "Avez-vous des questions supplémentaires ?",
   "account.order.subtitle": "La commande la plus récente apparaît en premier. Veuillez prévoir jusqu’à 5 minutes pour que les nouvelles commandes apparaissent ci-dessous.",
   "account.order.view_all_order.link": "Afficher toutes les commandes",
+  "account.order_history.filter.apply": "Trouver une commande",
+  "account.order_history.filter.clear": "Effacer tous les filtres",
+  "account.order_history.filter.hide": "Cacher tous les filtres",
+  "account.order_history.filter.label.date": "Période de temps",
+  "account.order_history.filter.label.order_no": "Rechercher un n° de commande",
```
Comment (mglatter) [inline src/assets/i18n/fr_FR.json:223]: Please check if it is OK to use the ° character or if this needs to be escaped.

File: src/assets/i18n/en_US.json [hunk 2]
```diff
@@ -210,11 +210,20 @@
...
   "account.notifications.table.product": "Product",
   "account.option.select.text": "Please select",
+  "account.order.load_more": "Load More",
   "account.order.most_recent.heading": "Most Recent Orders",
   "account.order.questions.note": "Please visit the <a href=\"{{0}}\">Help</a> area of our website for comprehensive order and shipping information or <a href=\"{{1}}\">Contact Us</a> 24 hours a day.",
   "account.order.questions.title": "Questions?",
   "account.order.subtitle": "The most recent order appears first. Please allow up to 5 minutes for new orders to appear below.",
   "account.order.view_all_order.link": "View All Orders",
+  "account.order_history.filter.apply": "Find Order",
+  "account.order_history.filter.clear": "Clear all filters",
```
[the remark this answers is not in the document]
Comment (mglatter) [reply src/assets/i18n/en_US.json]: Thanks for the hint. Done.

**Review by mglatter — DISMISSED** [review]

File: src/app/pages/account-order-history/account-order-filters/account-order-filters.component.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        },
+      },
+      {
+        fieldGroupClassName: 'row',
+        fieldGroup: [
+          {
+            className: 'col-12 col-md-6',
+            key: 'sku',
+            type: 'ish-text-input-field',
+            wrappers: ['form-field-vertical'],
```
Comment (SGrueber) [inline src/app/pages/account-order-history/account-order-filters/account-order-filters.component.ts:152]: can be removed,

File: src/app/pages/account-order-history/account-order-filters/account-order-filters.component.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            props: {
+              label: 'account.order_history.filter.label.sku',
```
Comment (SGrueber) [inline src/app/pages/account-order-history/account-order-filters/account-order-filters.component.ts:154]: set labelClass and fieldClass here

File: src/app/shared/formly/wrappers/vertical-wrapper/vertical-wrapper.component.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<div class="form-group row" [class.has-error]="showError" [attr.data-testing-id]="keyString + '-wrapper'">
```
Comment (SGrueber) [inline src/app/shared/formly/wrappers/vertical-wrapper/vertical-wrapper.component.html:1]: vertical wrapper can be removed

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/shared/formly/types/date-range-picker-field/date-range-picker-field.component.html [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      (input)="toDate = validateInput(toDate, dpToDate.value)"
+      [attr.aria-label]="'account.order_history.filter.label.date.aria_label' | translate"
+    />
+    <div class="input-group-append">
+      <div
+        class="btn btn-secondary"
+        (click)="d.toggle()"
+        (keydown.enter)="d.toggle()"
+        tabindex="-1"
+        title="Pick a date"
```
Comment (SGrueber) [inline src/app/shared/formly/types/date-range-picker-field/date-range-picker-field.component.html:78]: use a translation key

File: src/app/shared/formly/types/date-range-picker-field/date-range-picker-field.component.html [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        (input)="fromDate = validateInput(fromDate, dpFromDate.value)"
+        [attr.aria-label]="'account.order_history.filter.label.date.aria_label' | translate"
+      />
+      <div class="input-group-append">
+        <div
+          class="btn btn-secondary"
+          (click)="d.toggle()"
+          (keydown.enter)="d.toggle()"
+          tabindex="-1"
+          title="Pick a date"
```
Comment (SGrueber) [inline src/app/shared/formly/types/date-range-picker-field/date-range-picker-field.component.html:54]: translation key

**Review by SGrueber — CHANGES_REQUESTED** [review]
one more wish: if the user opens the pwa and has filter parameters in the url  - open the filter panel immediately:
http://pwa-gh-review-feat-order-list-filtering.azurewebsites.net/en/account/orders?from=2024-02-01&to=2024-02-29#order-list-top
otherwise the filter is used but he can`t see it

but we are almost there :-)

**Comment by DilaraGueler** [comment]
> one more wish: if the user opens the pwa and has filter parameters in the url - open the filter panel immediately: http://pwa-gh-review-feat-order-list-filtering.azurewebsites.net/en/account/orders?from=2024-02-01&to=2024-02-29#order-list-top otherwise the filter is used but he can`t see it
> 
> but we are almost there :-)

**Comment by iwiederhold** [comment]
Please check if the review server is available.

**Comment by M-Behr** [comment]
@DilaraGueler: Hi Dilara, it is possible to have two labels "Orders from" and "until" for date range picker (instead of time frame)? I think in this way it is more clear that these 2 input fields are used together.
@schadii and @mglatter would agree to this adjustment :)

**Comment by iwiederhold** [comment]
Hi Dilara, the filters differ slightly from the IAD. Is the Order Status filter not displayed?

**Comment by mglatter** [comment]
FR translation for added key will follow

**Comment by DilaraGueler** [comment]
> Hi Dilara, the filters differ slightly from the IAD. Is the Order Status filter not displayed?

Hi Inken, we had to remove the Order Status handling as long as it is not coming from ICM. If this is possible in the future, this part may be added.

**Comment by schadii** [comment]
Hi, would you mind considering keeping the filters after clicking on 'find order'? This way, users can easily see the filters they've applied and clear them if they wish.

File: src/assets/i18n/en_US.json [hunk 8]
```diff
@@ -210,11 +211,21 @@
...
   "account.option.select.text": "Please select",
+  "account.order.load_more": "Load more",
   "account.order.most_recent.heading": "Most Recent Orders",
   "account.order.questions.note": "Please visit the <a href=\"{{0}}\">Help</a> area of our website for comprehensive order and shipping information or <a href=\"{{1}}\">Contact Us</a> 24 hours a day.",
   "account.order.questions.title": "Questions?",
   "account.order.subtitle": "The most recent order appears first. Please allow up to 5 minutes for new orders to appear below.",
-  "account.order.view_all_order.link": "View All Orders",
+  "account.order.view_all_order.link": "View all orders",
+  "account.order_history.filter.apply": "Find order",
+  "account.order_history.filter.clear": "Clear all filters",
```
Comment (mglatter) [inline src/assets/i18n/en_US.json:221]: Changed back to sentence case as requested by Melanie

File: src/assets/i18n/de_DE.json [hunk 9]
```diff
@@ -153,6 +153,7 @@
   "account.customer.registered.title": "Vielen Dank für Ihre Registrierung!",
   "account.date.month": "Monat",
   "account.date.month.error.required": "Bitte wählen Sie einen Monat aus.",
+  "account.date.title": "Ein Datum wählen",
```
Comment (mglatter) [inline src/assets/i18n/de_DE.json:156]: Please check if the translation with infinitive form is fine for this title. We did the same for French. Please inform C3PO if any changes are needed.

**Review by mglatter — APPROVED** [review]

File: src/assets/i18n/de_DE.json (hunk 9 above)
Comment (DilaraGueler) [reply src/assets/i18n/de_DE.json:156]: @intershop/c3po-documentation-l10n 'Wählen Sie ein Datum aus' would probably be better at this point

**Review by schadii — APPROVED** [review]

**Review by iwiederhold — APPROVED** [review]
THX!

**Review by mglatter — APPROVED** [review]

File: src/app/pages/account-order-history/account-order-history-page.component.ts [hunk 10]
```diff
@@ -17,12 +18,58 @@ export class AccountOrderHistoryPageComponent implements OnInit {
...
+
+    this.orders$?.pipe(takeUntil(this.destroy$)).subscribe(orders => {
+      this.ordersAvailable = orders.length > 0;
+    });
+  }
+
+  /**
+   * Load orders
+   */
+  loadOrders(): void {
```
Comment (SGrueber) [inline src/app/pages/account-order-history/account-order-history-page.component.ts:44]: this method is not necessary and duplicates code: 
just call loadFilteredOrders({}) if you need the orders without a filter

**Review by SGrueber — CHANGES_REQUESTED** [review]

**Review by mglatter — DISMISSED** [review]

**Review by SGrueber — APPROVED** [review]

**Pull request merged**